### PR TITLE
test: Attach VM's XML and logs to a failed test

### PR DESCRIPTION
**Pull request by skobyda**
- [ ] Add these files to `test/common/link-patterns.json `

### CODE REVIEW

**Comment by martinpitt** [comment]
Fails everywhere:

    # test/machineslib.py:18:1: I001 [*] Import block is un-sorted or un-formatted

File: test/machineslib.py [hunk 1]
```diff
@@ -408,8 +410,44 @@ def setUp(self):
         # HACK: older c-ws versions always log an assertion, fixed in PR cockpit#16765
         self.allow_journal_messages("json_object_get_string_member: assertion 'node != NULL' failed")
 
+    def downloadVmXml(self, vm):
+        m = self.machine
+
+        m.execute(f"virsh dumpxml {vm} > /var/lib/cockpittest/pkg_specific_artifacts/{vm}.xml")
+        vms_xml = f"/var/lib/cockpittest/pkg_specific_artifacts/{vm}.xml"
```
Comment (martinpitt) [inline test/machineslib.py:417]: DRY, please use that variable in the m.execute() call.

File: test/machineslib.py [hunk 2]
```diff
@@ -408,8 +410,44 @@ def setUp(self):
...
+        dest_file = f"{self.label()}-{m.label}-{vm}.xml"
+        attach(vms_xml, move=False)
```
Comment (martinpitt) [inline test/machineslib.py:419]: Hmm, the first argument is supposed to be a file name, in the local test dir; vms_xml is the file name in the VM. How does that even work like that? You *first* need to m.download() the file and *then* attach it surely?

File: test/machineslib.py [hunk 3]
```diff
@@ -408,8 +410,44 @@ def setUp(self):
...
+    def downloadVmLog(self, vm):
+        m = self.machine
+
+        # Log file may not exist, if VM was never started
+        if m.execute(f"!test -f /var/log/libvirt/qemu/{vm}.log || echo exists") == "exists":
+            vms_log = f"/var/log/libvirt/qemu/{vm}.log"
+            dest_file = f"{self.label()}-{m.label}-{vm}.log"
+            attach(vms_log, move=True)
+            dest = os.path.abspath(dest_file)
+            m.download(vms_log, dest)
```
Comment (martinpitt) [inline test/machineslib.py:433]: Dito, this looks the wong way around. Also for repeating the vms_log path.

File: test/machineslib.py [hunk 4]
```diff
@@ -408,8 +410,44 @@ def setUp(self):
...
+        for vm in vms:
+            self.downloadVmXml(vm)
+            self.downloadVmLog(vm)
+
     def tearDown(self):
         b = self.browser
         if b.cdp.valid and b.is_present("#button.alert-link.more-button"):
             b.click("button.alert-link.more-button")
+
+        self.downloadVmsArtifacts()
```
Comment (martinpitt) [inline test/machineslib.py:451]: Please only do this on failure, otherwise this wastes too much storage space, and also test time.

**Review by martinpitt — CHANGES_REQUESTED** [review]
Thanks! This is a nice idea!

**Comment by skobyda** [comment]
Hmmm that's weird, locally I don't get the error  `test/machineslib.py:18:1: I001 [*] Import block is un-sorted or un-formatted`, even when running `test/static-code` manually

**Comment by martinpitt** [comment]
@skobyda ruff is installed in our cockpit/tasks container now, see https://github.com/cockpit-project/cockpit/blob/main/HACKING.md#setting-up-development-container

**Review by martinpitt — APPROVED** [review]
Nice, thanks! Do you have a failure somewhere to prove that it works?

**Comment by martinpitt** [comment]
[This run](https://cockpit-logs.us-east-1.linodeobjects.com/pull-1158-20230731-093210-86f4fc47-fedora-38-pybridge/log.html#64) has a flake, and the [result dir](https://cockpit-logs.us-east-1.linodeobjects.com/pull-1158-20230731-093210-86f4fc47-fedora-38-pybridge/index.html) at least has the XML (no log, though)

**Comment by martinpitt** [comment]
Also no log [here](https://cockpit-logs.us-east-1.linodeobjects.com/pull-1158-20230731-093210-86f4fc47-ubuntu-stable/index.html)

File: test/machineslib.py [hunk 5]
```diff
@@ -408,8 +412,50 @@ def setUp(self):
...
+        m.download(vms_xml, dest)
+        attach(dest, move=False)
+        print(f"Wrote {vm} XML to {dest_file}")
+
+    def downloadVmLog(self, vm):
+        m = self.machine
+
+        vms_log = f"/var/log/libvirt/qemu/{vm}.log"
+        # Log file may not exist, if VM was never started
+        if m.execute(f"!test -f {vms_log} || echo exists") == "exists":
```
Comment (martinpitt) [inline test/machineslib.py:431]: [This](https://cockpit-logs.us-east-1.linodeobjects.com/pull-1158-20230731-093210-86f4fc47-fedora-38-pybridge/log.html#64) shows why:

   bash: line 1: !test: command not found

/me buys a space

**Review by martinpitt — CHANGES_REQUESTED** [review]

**Review by martinpitt — APPROVED** [review]
Eeeeexcellent, thanks!

**Pull request merged**